presto-main/src/main/java/com/facebook/presto/operator/ArrayBlockEncodingBuffers.java

```diff
@@ -0,0 +1,226 @@
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(ArrayBlockEncodingBuffers.class).instanceSize();
+
+    // The buffer for the offsets for all incoming blocks so far
+    private byte[] offsetsBuffer;
+
+    // The address that the next offset value will be written to.
+    private int offsetsBufferIndex;
+
+    // This array holds the condensed offsets for each position for the incoming block.
+    private int[] offsets;
+
+    // The last offset in the offsets buffer
+    private int lastOffset;
+
+    // This array holds the offsets into its nested raw block for each top level row.
+    private int[] offsetsCopy;
+
+    private final BlockEncodingBuffers elementsBuffers;
+
+    public ArrayBlockEncodingBuffers(DecodedBlockNode decodedBlockNode)
+    {
+        elementsBuffers = createBlockEncodingBuffers(decodedBlockNode.getChildren().get(0));
+    }
+
+    @Override
+    public void resetBuffers()
+    {
+        bufferedPositionCount = 0;
+        offsetsBufferIndex = 0;
+        lastOffset = 0;
+        resetNullsBuffer();
+
+        elementsBuffers.resetBuffers();
+    }
+
+    @Override
+    protected void setupDecodedBlockAndMapPositions(DecodedBlockNode decodedBlockNode)
+    {
+        requireNonNull(decodedBlockNode, "decodedBlockNode is null");
+
+        decodedBlockNode = mapPositions(decodedBlockNode);
+        ColumnarArray columnarArray = (ColumnarArray) decodedBlockNode.getDecodedBlock();
+        decodedBlock = columnarArray.getNullCheckBlock();
+
+        populateNestedPositions(columnarArray);
+
+        elementsBuffers.setupDecodedBlockAndMapPositions(decodedBlockNode.getChildren().get(0));
+    }
+
+    @Override
+    public void accumulateRowSizes(int[] serializedRowSizes)
+    {
+        for (int i = 0; i < positionCount; i++) {
+            serializedRowSizes[i] += POSITION_SIZE;
+        }
+
+        offsetsCopy = ensureCapacity(offsetsCopy, positionCount + 1);
+        System.arraycopy(offsets, 0, offsetsCopy, 0, positionCount + 1);
+
+        elementsBuffers.accumulateRowSizes(offsetsCopy, positionCount, serializedRowSizes);
+    }
+
+    @Override
+    protected void accumulateRowSizes(int[] positionOffsets, int positionCount, int[] rowSizes)
+    {
+        if (this.positionCount == 0) {
+            return;
+        }
+
+        int lastOffset = positionOffsets[0];
+        for (int i = 0; i < positionCount; i++) {
+            int offset = positionOffsets[i + 1];
+            rowSizes[i] += POSITION_SIZE * (offset - lastOffset);
+            lastOffset = offset;
+            positionOffsets[i + 1] = offsets[offset];
+        }
+
+        elementsBuffers.accumulateRowSizes(positionOffsets, positionCount, rowSizes);
+    }
+
+    @Override
+    public void setNextBatch(int positionsOffset, int batchSize)
+    {
+        this.positionsOffset = positionsOffset;
+        this.batchSize = batchSize;
+
+        // The nested level positionCount could be 0.
```

Comment:
> I don't quite understand this comment. What does "nested level" mean? Can you elaborate a bit?
> 
> In general I found "top-level" and "nested-level" often can often be a bit difficult to understand without more context / examples.

By "Nested level" I meant the child block/blockEncodingBuffers as to the composite block/blockEncodingBuffers, e.g. the values Block to an ArrayBlock, and valuesBuffers to the ArrayBlockEncodingBuffers.  How about the following comments? Is it "child level" easier to understand? If so I will change it to "child level"
```
         // If all positions for the ArrayBlock to be copied are null, the number of positions to copy for its 
        // nested values block could be 0. In such case we don't need to proceed.
        if (this.positionCount == 0) {
            return;
        }
```
Also modified the comments of offsetsCopy and valuesBuffers
```
    // This array holds the offsets into its nested values block for each row in the ArrayBlock.
    private int[] offsetsCopy;

    // The BlockEncodingBuffers for the nested values Block of the ArrayBlock
    private final BlockEncodingBuffers valuesBuffers;
```

Reply:
@yingsu00 :

I see. So the "nested value block" it's basically the underlying value block of a structural block (e.g. ArrayBlock). For an `ARRAY<BIGINT>`, the `valueBlock` type is `LongArrayBlock`.

`Nested values block` used in your second comment looks good to me. -- It's not so easy to understand since due to it's closely related to the low-level detail of structural data representation. So let's just try our best to explain :)